[Peter] Chapter 5. JPA 활용

devgeon opened the pull request:
<!--
  제목은 `[닉네임] 챕터 제목` 로 작성해 주세요.
  예시: [Angela] Chapter 1. 서버란 무엇인가(소켓&멀티 프로세스)
-->

## ✅ 워크북 체크리스트

- [x] 모든 핵심 키워드 정리를 마쳤나요?
- [x] 핵심 키워드에 대해 완벽히 이해하셨나요?
- [x] 이론 학습 이후 직접 실습을 해보는 시간을 가졌나요?
- [x] 미션을 수행하셨나요?
- [x] 미션을 기록하셨나요?
<br>

## ✅ 컨벤션 체크리스트

- [x] 디렉토리 구조 컨벤션을 잘 지켰나요?
- [x] pr 제목을 컨벤션에 맞게 작성하였나요?
- [x] pr에 해당되는 이슈를 연결하였나요?
- [x] 적절한 라벨을 설정하였나요?
- [ ] 스터디원들에게 code review를 요청하기 위해 reviewer를 등록하였나요?
- [x] 닉네임/main 브랜치의 최신 상태를 반영하고 있는지 확인했나요?
<br>

## 📌 주안점

<!--
  (Optional)
  리뷰 시에 유심히 봐주었으면 하는 부분을 설명합니다.
-->

devgeon committed:
fix/04 - 일부 entity의 @Builder.Default 누락된 필드 수정

devgeon committed:
fix/05 - 상세 주소에 NULL 허용

devgeon committed:
test/05 - JPQL 쿼리 테스트 코드 작성

devgeon committed:
mission/05 - JPQL 쿼리 메서드 작성

ijnuegnaj reviewed (approved):
수고하셨습니다!

Yooze0x4c8 reviewed (approved):
수고하셨습니다

taebin2 reviewed (approved):
고생하셨습니다~

Dimo-2562 commented on src/main/java/com/umc9th/peter/domain/member/repository/MemberMissionRepository.java:
```diff
@@ -10,6 +13,8 @@
 @Repository
 public interface MemberMissionRepository extends JpaRepository<MemberMission, Long> {
 
+    Page<MemberMission> findByMemberIdAndStatus(Long memberId, MissionStatus status, Pageable pageable);
```
단순히 유저가 진행중인 미션의 정보뿐만 아니라 미션의 정보와 미션이 수행된 가게의 정보도 필요하니까 조인이 아마 수행될겁니다. 단순히 조인하면 쿼리가 많이 나가므로 이것도 Query로 fetch join 해주시는 게 좋을 거 같습니다.

```
-- 1. MemberMission 페이징 조회
SELECT * FROM user_mission 
WHERE member_id = ? AND status = ?
LIMIT 10;

-- 2. 각 UserMission마다 Mission 조회 (N+1)
SELECT * FROM mission WHERE mission_id = ?;  -- 1번
SELECT * FROM mission WHERE mission_id = ?;  -- 2번
... (10번)

-- 3. 각 Mission마다 Store 조회 (N+1)
SELECT * FROM store WHERE store_id = ?;  -- 1번
SELECT * FROM store WHERE store_id = ?;  -- 2번
... (10번)
```

Dimo-2562 replied:
저는 Page 객체에 쓸데없는 정보가 너무 많은거 같아서 그냥 List로 조회해온뒤에 추가 정보는 따로 생성해서 반환해주는 편입니다.

그리고 초기 조회 메서드랑 페이징 메서드를 분리하는 식으로 구현했었습니다.
아래는 채팅방에서 메시지 가져오는 코드인데 참고해주시면 좋을 거 같습니당

```
Pageable pageable = PageRequest.of(0, size + 1);
List<ChatMessage> messages = findMessagesWithCursor(roomId, cursor, pageable);

boolean hasNext = messages.size() > size;
if (hasNext) {
      messages = messages.subList(0, size);
}

Long nextCursor = hasNext && !messages.isEmpty()
                ? messages.get(0).getId() : null;
```

```
    @Query("""
            SELECT m FROM ChatMessage m
            JOIN FETCH m.sender
            LEFT JOIN FETCH m.chatMessageImgs
            WHERE m.chatRoom.id = :chatRoomId
            AND m.id < :cursor
            ORDER BY m.createdAt DESC
            """)
    List<ChatMessage> findByRoomIdAndIdLessThanOrderByCreatedAtDesc(
            @Param("chatRoomId") Long chatRoomId,
            @Param("cursor") Long cursor,
            Pageable pageable);
```

devgeon replied:
Fetch join과 paging을 같이 사용하면 모든 데이터를 가져와서 메모리에서 paging을 시도하므로 OOM의 위험이 있습니다.
말씀하신 쿼리를 추가하면서 fetch join 대신 default_batch_fetch_size를 적용하여 N+1 문제를 해결하도록 작성했습니다.

devgeon replied:
List로 받아오는 방법도 좋은 것 같습니다.
좋은 의견 감사합니다!

Dimo-2562 commented on src/main/java/com/umc9th/peter/domain/mission/repository/MissionRepository.java:
```diff
@@ -0,0 +1,21 @@
+            "JOIN FETCH m.missionDistrictList md " +
+            "WHERE md.district.id  = :districtId " +
+            "AND :now BETWEEN m.beginAt AND m.endAt")
+    Page<Mission> findAvailableMissionsByDistrictId(Long districtId, LocalDateTime now, Pageable pageable);
```
진행 가능한 미션 목록 조회할 때 미션과 관련된 store 정보 또한 조인이 되어야 미션을 건 가게의 정보도 알 수 있을거 같아요!

그리고 홈 화면 위쪽 탭에 있는 내가 이 지역에서 진행한 미션의 개수, 현재 포인트, 지역 정보 조회하는 쿼리도 해보시면 좋을 거 같습니다

devgeon replied:
앞선 경우와 마찬가지로 fetch join 대신 default_batch_fetch_size를 적용했습니다.

devgeon replied:
특정 멤버가 지역별로 완료한 미션의 개수는 멤버가 등록한 지역을 N:M으로 매핑하는 테이블에 완료한 미션 개수를 카운트하는 컬럼을 추가하는 식으로 작성했습니다.

홈화면에 표시되는 정보라서 빈번하게 조회되기도 하고, 미션 완료 즉시 값이 갱신되어야 하므로 위와 같이 별도 컬럼에 기록하는 것이 좋다고 생각했습니다.

해당 값은 서비스 레이어에서 미션을 완료할 때 증가시키는 식으로 관리합니다.

Dimo-2562 reviewed (commented):
테스트 코드까지 수고 많으셨습니다!

devgeon committed:
mission/05 - mission 관련 쿼리에 fetch join 적용

devgeon committed:
fix/05 - 쿼리 파라미터명 명시적으로 작성

devgeon committed:
fix/05 - wrapper class로 통일

devgeon committed:
refact/05 - repository 리턴값에 optional 적용

devgeon committed:
fix/05 - findAvailableMissionsByDistrictId 쿼리 수정

- fetch join 대신 batch size 적용 (paging으로 인한 OOM 예방)
- 해당 유저가 기존에 수락, 완료, 거절한 미션은 표시되지 않도록 변경

devgeon committed:
mission/05 - MemberDistrict 엔티티 추가 (멤버별 지역, 완료 미션 수 기록)

devgeon committed:
mission/05 - 홈 화면 상단부 쿼리 작성
